### ISSUE TITLE
Hotkey sometimes lets original keypress through

### ISSUE THREAD

**Issue by AndreasArvidsson** [talonvoice/talon#458]
I use is a foot switch button as a repeater. The foot switch sends the keypress `keypad_3`. All of these works fine but sometimes the last command will get repeated AND the number `3` will get inserted. This only seems to happen when I use vscode and specifically commands that required a command server. I'm thinking this might have something to do that the command client issues `key('ctrl-shift-f17')`.

os: windows10

```
core | action core.replace_command([])
user\andreas\misc\foot_switch\foot_switch.talon | action core.run_hotkey(ScriptImpl("keypad_3"))
user\andreas\misc\foot_switch\foot_switch.talon | action core.run_talon_script(ResourceContext("user.andreas.misc.foot_switch.foot_switch.talon"), TalonScript(code='user.foot_switch_key(3)'), ['keypad_3'])
user\andreas\misc\foot_switch\foot_switch.talon | action user.foot_switch_key(3)
user\andreas\misc\foot_switch\foot_switch.talon | action core.run_hotkey(ScriptImpl("keypad_3"))
user\andreas\misc\foot_switch\foot_switch.talon | action core.run_talon_script(ResourceContext("user.andreas.misc.foot_switch.foot_switch.talon"), TalonScript(code='user.foot_switch_key(3)'), ['keypad_3'])
user\andreas\misc\foot_switch\foot_switch.talon | action user.foot_switch_key(3)
user\andreas\misc\foot_switch\foot_switch.py | action user.foot_switch_right_down()
user\andreas\misc\foot_switch\foot_switch.py | action core.repeat_command(1)
core | action core.run_command(CommandImpl(Rule("comment")), C(__comment__, comment))
core | action core.run_talon_script(ResourceContext("user.andreas.apps.vscode.vscode.talon"), TalonScript(code='code.toggle_comment()'), C(__comment__, comment))
user\andreas\apps\vscode\vscode.talon | action code.toggle_comment()
user\andreas\apps\vscode\vscode.py | action user.vscode('editor.action.commentLine')
user\andreas\apps\vscode\command_client.py | action user.trigger_command_server_command_execution()
user\andreas\apps\vscode\command_client.py | action key('ctrl-shift-f17')
user\andreas\misc\keys\keys.py | action key('ctrl-shift-f17')
core | action core.replace_command([])

```

**Comment by lunixbochs** [comment]
Can you run some external key logger to see the sequence of keypresses?

**Comment by AndreasArvidsson** [comment]
I would love to but it seems I can't get a single keylogger to work. They either get quarantine by my virus protection for containing trojans and the few that actually managed to install correctly don't show anything in the logs. I'm going to have to give it some more time another day.

**Comment by lunixbochs** [comment]
try spy++, it's a microsoft tool

**Comment by AndreasArvidsson** [comment]
I finally got spy++ working. I had to start 64bit version. The one launched from inside visual studio showed no messages. Anyway here is the last bit of the log when the error occurred.
```

<000873> 000000000B070604 P WM_KEYUP nVirtKey:VK_NUMPAD3 cRepeat:1 ScanCode:51 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000874> 000000000B070604 P WM_KEYUP nVirtKey:VK_NUMPAD3 cRepeat:1 ScanCode:51 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000875> 000000000B070604 P WM_KEYDOWN nVirtKey:VK_SHIFT cRepeat:1 ScanCode:2A fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000876> 000000000B070604 P WM_KEYDOWN nVirtKey:VK_CONTROL cRepeat:1 ScanCode:1D fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000877> 000000000B070604 P WM_KEYDOWN nVirtKey:VK_F17 cRepeat:1 ScanCode:68 fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000878> 000000000B070604 P WM_KEYUP nVirtKey:VK_F17 cRepeat:1 ScanCode:68 fExtended:0 fAltDown:0 fRepeat:1 fUp:1
<000879> 000000000B070604 P WM_KEYUP nVirtKey:VK_CONTROL cRepeat:1 ScanCode:1D fExtended:0 fAltDown:0 fRepeat:1 fUp:1
<000880> 000000000B070604 P WM_KEYUP nVirtKey:VK_SHIFT cRepeat:1 ScanCode:2A fExtended:0 fAltDown:0 fRepeat:1 fUp:1
<000881> 000000000B070604 P WM_KEYUP nVirtKey:VK_NUMPAD3 cRepeat:1 ScanCode:51 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000882> 000000000B070604 P WM_KEYUP nVirtKey:VK_NUMPAD3 cRepeat:1 ScanCode:51 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000883> 000000000B070604 P WM_KEYDOWN nVirtKey:VK_SHIFT cRepeat:1 ScanCode:2A fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000884> 000000000B070604 P WM_KEYDOWN nVirtKey:VK_CONTROL cRepeat:1 ScanCode:1D fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000885> 000000000B070604 P WM_KEYDOWN nVirtKey:VK_F17 cRepeat:1 ScanCode:68 fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000886> 000000000B070604 P WM_KEYUP nVirtKey:VK_F17 cRepeat:1 ScanCode:68 fExtended:0 fAltDown:0 fRepeat:1 fUp:1
<000887> 000000000B070604 P WM_KEYUP nVirtKey:VK_CONTROL cRepeat:1 ScanCode:1D fExtended:0 fAltDown:0 fRepeat:1 fUp:1
<000888> 000000000B070604 P WM_KEYUP nVirtKey:VK_SHIFT cRepeat:1 ScanCode:2A fExtended:0 fAltDown:0 fRepeat:1 fUp:1
<000889> 000000000B070604 P WM_KEYUP nVirtKey:VK_NUMPAD3 cRepeat:1 ScanCode:51 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000890> 000000000B070604 P WM_KEYDOWN nVirtKey:VK_NUMPAD3 cRepeat:1 ScanCode:51 fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000891> 000000000B070604 P WM_CHAR chCharCode:'51' (51) cRepeat:1 ScanCode:51 fExtended:0 fAltDown:0 fRepeat:0 fUp:0
```

**Comment by lunixbochs** [comment]
Yeah this might be that you're not actually pressing the exact hotkey due to the vscode sync pressing modifier keys.

If you bind key(3), then stuff like ctrl-3 will still go through. I don't think this is a talon bug, it's probably the vscode rpc messing up global keyboard flag state.

**Comment by AndreasArvidsson** [comment]
Yes could be that. Looking in the log of spy++ I can see one pattern. When there is no problem the foot switch sends 2 WM_KEYUP events, but when there is a problem the second event turns into WM_KEYDOWN.

**Comment by AndreasArvidsson** [comment]
I just tried to add hotkeys for the modifiers to see if I could catch them but to no avail. Still a single 3 was written. 
```
key(ctrl-keypad_3):
    print("ctrl-keypad_3")
key(shift-keypad_3):
    print("shift-keypad_3")
key(ctrl-shift-keypad_3):
    print("ctrl-shift-keypad_3")
```

**Comment by lunixbochs** [comment]
It never happens outside vscode right? It really must be caused by the overlapping keypresses.

**Comment by AndreasArvidsson** [comment]
I think it might have happened a few times outside but in vscode it happens quite frequently. Also doing commands in vscode that don't require the command server don't have the same error frequency. The overlapping keypresses is a likely suspect.

**Comment by AndreasArvidsson** [comment]
I've kept an eye on this the last day and I definitely have had numbers being inserted by the foots which not in vscode. Several times in the terminal for example. Here is a log from the terminal where a 0 was inserted. Same behavior as we saw in vscode. When a WM_KEYDOWN is registered the key event gets through.
```
<000025> 00000000000105D8 P WM_KEYUP nVirtKey:VK_NUMPAD0 cRepeat:1 ScanCode:52 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000026> 00000000000105D8 P WM_KEYUP nVirtKey:VK_NUMPAD0 cRepeat:1 ScanCode:52 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000027> 00000000000105D8 P WM_KEYUP nVirtKey:VK_NUMPAD0 cRepeat:1 ScanCode:52 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000028> 00000000000105D8 P WM_KEYUP nVirtKey:VK_NUMPAD0 cRepeat:1 ScanCode:52 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000029> 00000000000105D8 P WM_KEYUP nVirtKey:VK_NUMPAD0 cRepeat:1 ScanCode:52 fExtended:0 fAltDown:0 fRepeat:0 fUp:1
<000030> 00000000000105D8 P WM_KEYDOWN nVirtKey:VK_NUMPAD0 cRepeat:1 ScanCode:52 fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000031> 00000000000105D8 P WM_CHAR chCharCode:'48' (48) cRepeat:1 ScanCode:52 fExtended:0 fAltDown:0 fRepeat:0 fUp:0
<000032> 00000000000105D8 P WM_KEYUP nVirtKey:VK_NUMPAD0 cRepeat:1 ScanCode:52 fExtended:0 fAltDown:0 fRepeat:1 fUp:1
```

**Comment by lunixbochs** [comment]
Wait are you holding down the pedal with key repeat on? We release the hotkey temporarily while your script is running so your script can press the key.

**Comment by AndreasArvidsson** [comment]
Key repeat should not be on but both downpress and release sends the key event. That's the only way I can do a hold action until talon supports key:up.

**Comment by lunixbochs** [comment]
Can you make pedal press and release use two different keys? That way the pedal release won't trigger the same key while talon is processing your hotkey

A compromise on my end could be to hook actions.key() to remove the hotkey temporarily anytime you try to press the key from talon

**Comment by AndreasArvidsson** [comment]
Unfortunately not. I have three settings on my foot switch
1. Single press
2. Double press. Same key on press and release
3. Hold. Continuously sends the same key

**Comment by lunixbochs** [comment]
this should be fixed in 0.2.0-489
